frame/uniques/src/lib.rs

```diff
@@ -0,0 +1,1182 @@
+			}
+
+			Asset::<T, I>::insert(&class, &instance, &details);
+			Self::deposit_event(Event::ApprovalCancelled(class, instance, origin, old));
+
+			Ok(())
+		}
+
+		/// Cancel the prior approval for the transfer of an asset by a delegate.
+		///
+		/// Origin must be either ForceOrigin or Signed origin with the signer being the Admin
+		/// account of the asset `class`.
+		///
+		/// - `class`: The class of the asset of whose approval will be cancelled.
+		/// - `instance`: The instance of the asset of whose approval will be cancelled.
+		/// - `maybe_check_delegate`: If `Some` will ensure that the given account is the one to
+		///   which permission of transfer is delegated.
+		///
+		/// Emits `ApprovalCancelled` on success.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::force_cancel_approval())]
+		pub(super) fn force_cancel_approval(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			#[pallet::compact] instance: T::InstanceId,
+			maybe_check_delegate: Option<<T::Lookup as StaticLookup>::Source>,
+		) -> DispatchResult {
+			T::ForceOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(|origin| -> DispatchResult {
+					let origin = ensure_signed(origin)?;
+					let d = Class::<T, I>::get(&class).ok_or(Error::<T, I>::Unknown)?;
+					ensure!(&origin == &d.admin, Error::<T, I>::NoPermission);
+					Ok(())
+				})?;
+
+			let maybe_check_delegate = maybe_check_delegate.map(T::Lookup::lookup).transpose()?;
+
+			let mut details = Asset::<T, I>::get(&class, &instance)
+				.ok_or(Error::<T, I>::Unknown)?;
+			let old = details.approved.take().ok_or(Error::<T, I>::NoDelegate)?;
+			if let Some(check_delegate) = maybe_check_delegate {
+				ensure!(check_delegate == old, Error::<T, I>::WrongDelegate);
+			}
+
+			Asset::<T, I>::insert(&class, &instance, &details);
+			Self::deposit_event(Event::ApprovalCancelled(class, instance, details.owner, old));
+
+			Ok(())
+		}
+
+		/// Transfer some asset instance from a target account to a destination account by an
+		/// approved delegate.
+		///
+		/// Origin must be Signed and there must be an approval in place for the signer on this
+		/// asset `instance`.
+		///
+		/// - `class`: The class of the asset to transfer.
+		/// - `instance`: The instance of the asset to transfer.
+		/// - `dest`: The account to which ownership of the asset instance will be transferred.
+		///
+		/// Emits `TransferredApproved` on success.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::transfer_approved())]
+		pub(super) fn transfer_approved(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			#[pallet::compact] instance: T::InstanceId,
+			dest: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			let dest = T::Lookup::lookup(dest)?;
+
+			let mut details = Asset::<T, I>::get(&class, &instance).ok_or(Error::<T, I>::Unknown)?;
```

Comment:
Could check `details.owner == dest` as an optimization?

Reply:
Surprised that you havent yet ripped out a `do_transfer` function shared across pallets with such optimizations

Reply:
As I already said, there's no point in making optimisations that are not recognised by the weights system. It complicates the code paths and doesn't lead to greater transaction volume.

Reply:
The `transfer` code is sufficiently small and different between uses that I saw little point at this stage.